clientconn.go

```diff
@@ -1331,19 +1331,32 @@ func (ac *addrConn) createTransport(addr resolver.Address, copts transport.Conne
 	case <-prefaceReceived.Done():
 		// We got the preface - huzzah! things are good.
 		ac.mu.Lock()
-		defer ac.mu.Unlock()
-		defer prefaceReceived.Fire()
 		if connClosed.HasFired() {
 			// onClose called first; go idle but do nothing else.
 			if ac.state != connectivity.Shutdown {
 				ac.updateConnectivityState(connectivity.Idle, nil)
 			}
+			ac.mu.Unlock()
 			return nil
 		}
+		if ac.state == connectivity.Shutdown {
+			// This can happen if the subConn was removed while in `Connecting`
+			// state. tearDown() would have set the state to `Shutdown`, but
+			// would not have closed the transport since ac.transport would not
+			// been set at that point.
+			err := fmt.Errorf("subConn closed before receiving preface")
+			// We have to unlock here because newTr.Close() calls onClose()
+			// inline, which requires locking ac.mu.
+			ac.mu.Unlock()
+			newTr.Close(err)
+			channelz.Warningf(logger, ac.channelzID, "grpc: addrConn.createTransport failed to connect to %v: %v", addr, err)
```

Comment:
I think this log message is misleading.

Isn't there already a channelz log entry for the subchannel being deleted?

This path literally means we _did_ connect to the address successfully, but we are killing it because it was deleted.

Reply:
Yes, we have a log entry for the subChannel being deleted. But that would just say something like:
```
Subchanel(id:3) deleted
```

Do you think the message is better now? We are still going to say `addrConn.createTransport failed to connect`, but we will say why though, which is `subConn shutdown initiated by balancer before receiving server preface`.